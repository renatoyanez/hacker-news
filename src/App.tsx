import { ReactElement, useState, useEffect } from 'react';
import Routes from './routes';
import { FaveNewsContext } from 'context/favesContext';
import { Layout } from 'components';
import { IHits } from 'interfaces';

const App = (): ReactElement => {
	// Get local storage in the first place
	const getFavesLocalStorage = JSON.parse(
		localStorage.getItem('faves') as string,
	) as IHits[];

	// Set local storage after first render if it's not already
	useEffect(() => {
		if (!getFavesLocalStorage) {
			localStorage.setItem('faves', JSON.stringify([]));
		}
	}, []);

	const [favedHits, setFavedHits] = useState<IHits[]>(
		getFavesLocalStorage || ([] as IHits[]),
	);

	const handleFavedHits = (hit: IHits, faves: IHits[]) => {
		const found =
			faves &&
			faves.length &&
			faves.length > 0 &&
			faves.find(item => item.objectID === hit.objectID);

		if (found) {
			// Take it out if it exists
			const filtered =
				faves && faves.length && faves.length > 0
					? faves.filter(item => item.objectID !== found.objectID)
					: ([] as IHits[]);
			setFavedHits([...filtered]);
			localStorage.setItem('faves', JSON.stringify([...filtered]));
		} else {
			// Add it if it doesn't
			setFavedHits([...faves, hit]);
			localStorage.setItem('faves', JSON.stringify([...faves, hit]));
		}
	};
	return (
		<div>
			<Layout>
				<FaveNewsContext.Provider
					value={{
						favedHits,
						handleFavedHits: (hit: IHits, faves: IHits[]) => {
							handleFavedHits(hit, faves);
						},
					}}
				>
					<Routes />
				</FaveNewsContext.Provider>
			</Layout>
		</div>
	);
};

export default App;
