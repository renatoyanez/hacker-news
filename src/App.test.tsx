import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
// import { render, screen } from '@testing-library/react';
import App from './App';

describe('App test', () => {
	it('App renders correctly', () => {
		const jsx = (
			<BrowserRouter>
				<App />
			</BrowserRouter>
		);
		const renderedComponent = renderer.create(jsx).toJSON();
		expect(renderedComponent).toMatchSnapshot();
	});
});
