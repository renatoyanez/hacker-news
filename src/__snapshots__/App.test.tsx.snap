// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test App renders correctly 1`] = `
<div>
  <div>
    <div
      className="header"
    >
      <p
        className="textStyle"
      >
        HACKER NEWS
      </p>
    </div>
    <div
      className="layout"
    >
      <div
        style={
          Object {
            "marginBottom": "4.375rem ",
          }
        }
      >
        <div>
          <button
            className="selectedButton first"
            onClick={[Function]}
          >
            <p
              className="textStyle2"
            >
              All
            </p>
          </button>
          <button
            className="button second"
            onClick={[Function]}
          >
            <p
              className="textStyle5"
            >
              My faves
            </p>
          </button>
        </div>
      </div>
      <div
        className="home"
      >
        <div>
          <div
            className="container"
          >
            <div
              className="selectedText"
              data-testid="dropdown-component"
              onClick={[Function]}
            >
              <p
                className="textStyle4"
              >
                Select your news
              </p>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "marginBottom": "2.375rem",
            }
          }
        />
        <div>
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
